### PR TITLE
Release/0.2.0

### PR DESCRIPTION
**Pull request by oguzhanunlu**
- [x] Fix SchemaVer.parse_schemaver parses invalid SchemaVer #7 

- [x] Update README markdown in according with CommonMark #13 

- [x] Fix invalid cache-lookup #11 

- [x] Add retry-policy #10 

- [x] Add configurable cache max-age for schemas #6 

- [x] Switch to rubygems provider #5 

- [x] Remove Ruby 2.0.0 support #15 

- [x] Remove unnecessary ResolverError check from cache #16 

- [x] Bump version in README (https://github.com/snowplow/iglu-ruby-client/blame/master/README.md#L47)

- [x] Add CHANGELOG

### CODE REVIEW

**Comment by alexanderdean** [comment]
Missing space in commit message "Add configurable cache max-age for schemas (close#6)"

**Comment by alexanderdean** [comment]
"Bump release-image version to 0.2.0 in README" belongs in the "Prepare for release" commit

**Comment by oguzhanunlu** [comment]
Thanks @alexanderdean , just fixed them.

File: .travis.yml [hunk 1]
```diff
@@ -15,12 +15,13 @@ install:
...
-  on: 
-    tags: true
-    condition: '"$($TRAVIS_BUILD_DIR/.travis/is_valid_tag.rb $TRAVIS_TAG)" == "" && $? == 0'
+deploy:
+  provider: rubygems
+  api_key: $RUBYGEMS_SNOWPLOW_API_KEY
+  gemspec: iglu-ruby-client.gemspec
+  on:
+   tags: true
+
```
Comment (chuwy) [inline .travis.yml]: Please, add a `condition` from previous version as well (and `.travis/is_valid_tag.rb` as it is required), I should have warn you about it. 

Without `condition` pushing could result in a quite serious issue when we would push `0.3.0.beta` tag, but `0.3.0` by mistake remains in source code - as a result `0.3.0` will be published to RubyGems, which is in fact should have been `0.3.0.beta` (by our intentions, e.g. git tag). And we cannot delete releases from RubyGems, which means we'll leave invalid release on RubyGems.

We have this situation right now: `0.2.0.beta` is on RubyGems, but in source code there's `0.2.0.beta.2`. It's quite ok for beta-releases, but should never happen for final releases. `condition` makes us safe here.

File: lib/iglu-client/resolver.rb [hunk 2]
```diff
@@ -16,11 +16,12 @@ module Iglu
 
   # Iglu Client. Able to fetch schemas only from Iglu Central
   class Resolver
-    attr_reader :registries, :cache
+    attr_reader :registries, :cache, :cache_ttl
 
     def initialize(registries)
       @registries = registries.unshift(Registries.bootstrap)
       @cache = Hash.new
+      @cache_ttl = 1 # hour
```
Comment (chuwy) [inline lib/iglu-client/resolver.rb]: `cacheTtl` by no means should be in hours - this is always seconds.

File: lib/iglu-client/version.rb [hunk 3]
```diff
@@ -10,5 +10,5 @@
 # See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 
 module Iglu
-  CLIENT_VERSION = '0.1.0'
+  CLIENT_VERSION = '0.2.0.beta.2'
```
Comment (chuwy) [inline lib/iglu-client/version.rb]: As I mentioned in first comment - there's only `0.2.0.beta` published on RubyGems. Version in latest tag and asset on RubyGems should always be compatible.

File: lib/iglu-client/resolver.rb [hunk 4]
```diff
@@ -16,11 +16,12 @@ module Iglu
 
   # Iglu Client. Able to fetch schemas only from Iglu Central
   class Resolver
-    attr_reader :registries, :cache
+    attr_reader :registries, :cache, :cache_ttl
```
Comment (chuwy) [inline lib/iglu-client/resolver.rb]: I expect to see test coverage for this new feature with **at least** following cases:

* Parse resolver config JSON with `cacheTtl`
* Invalidate schema after some period of time (make sure new fetch attempt is undertaken). You can use `sleep` here
* Confirm that no new fetch attempts are undertaken after small enough period

**Review by chuwy — CHANGES_REQUESTED** [review]
Hey @oguzhanunlu,

This is a good start! I left few comments - ones regarding new `cacheTtl` are most important - Iglu Scala Client uses seconds as TTL and we cannot have clients with inconsistent configuration. Also, tests are very important.

**Comment by chuwy** [comment]
Also, regarding tests - I would expect we can parse configuration with both:

1. `cacheTtl: null`
2. missing `cacheTtl`

These are both valid `resolver_config/1-0-2`, but might have different representations in Ruby.

**Comment by oguzhanunlu** [comment]
@chuwy I think 0.2.0.beta.3 meets your review except a test confirming that only one fetch operation is performed before cacheTtl expires. I spent too much time unnecessarily for that test and wanted to release 0.2.0.beta.3 for your review.

File: lib/iglu-client/resolver.rb [hunk 5]
```diff
@@ -16,11 +16,12 @@ module Iglu
...
   # Iglu Client. Able to fetch schemas only from Iglu Central
   class Resolver
-    attr_reader :registries, :cache
+    attr_reader :registries, :cache, :cacheTtl
 
-    def initialize(registries)
+    def initialize(registries, cacheTtl=60)
       @registries = registries.unshift(Registries.bootstrap)
       @cache = Hash.new
+      @cacheTtl = cacheTtl.nil? ? 60 : cacheTtl
```
Comment (chuwy) [inline lib/iglu-client/resolver.rb]: I don't think we can afford default values here. If `cacheTtl` isn't set then our cache simply does not expire and lives in application until it is terminated.

File: lib/iglu-client/resolver.rb [hunk 6]
```diff
@@ -50,14 +52,24 @@ def lookup_schema(schema_key)
...
         else
-          cache_result
+          if cache_result.is_a? (Array)
+              store_time = cache_result[1]
+              time_diff = (fetch_time - store_time).round
+              if time_diff >= @cacheTtl
+                 @cache.delete(schema_key)
+               end
+               cache_result[0]
+          else # backward compliance
```
Comment (chuwy) [inline lib/iglu-client/resolver.rb]: Not sure what does `backward compliance` mean here. Are we returning `[schema, time]` from `lookup_schema` function? Cache mechanics should be a part of library internals and user never cares of timestamp, therefore `lookup_schema` should always return only `schema`.

File: lib/iglu-client/resolver.rb [hunk 7]
```diff
@@ -50,14 +52,24 @@ def lookup_schema(schema_key)
...
         if cache_result.is_a?(Registries::ResolverError)
           raise cache_result
         else
-          cache_result
+          if cache_result.is_a? (Array)
+              store_time = cache_result[1]
+              time_diff = (fetch_time - store_time).round
+              if time_diff >= @cacheTtl
+                 @cache.delete(schema_key)
+               end
```
Comment (chuwy) [inline lib/iglu-client/resolver.rb]: Broken indentation.

File: lib/iglu-client/resolver.rb (hunk 6 above)
Comment (chuwy) [reply lib/iglu-client/resolver.rb]: Also, please make return type clear in function's description.

File: lib/iglu-client/resolver.rb [hunk 8]
```diff
@@ -67,9 +79,9 @@ def self.parse(json)
       schema = Registries.bootstrap.lookup_schema(schema_key)
       data = get_data(json)
       if JSON::Validator.validate!(schema, data)
-        # cache_size = data[:cacheSize]
```
Comment (chuwy) [inline lib/iglu-client/resolver.rb:70]: No commented code should be in production code.

File: spec/iglu-client/resolver_spec.rb [hunk 9]
```diff
@@ -22,14 +22,48 @@
...
     resolver = Iglu::Resolver.parse(@json_config)
-    expect(resolver.registries.length) == 2
-    expect(resolver.registries[1].config.name) == "Iglu Central"
+    expect(resolver.registries.size).to eq(2)
+    expect(resolver.cacheTtl).to eq(60)
+    expect(resolver.registries[1].config.name).to eq("Iglu Central")
   end
-  
+
+  it 'correctly parses a standard resolver configuration containing cacheTtl (not null)' do
```
Comment (chuwy) [inline spec/iglu-client/resolver_spec.rb:40]: This is amount of unit-tests I expected for this feature. Very good.

File: spec/iglu-client/resolver_spec.rb [hunk 10]
```diff
@@ -22,14 +22,48 @@
...
+  it 'properly invalidates resolver cache per cacheTtl' do
+    resolver = Iglu::Resolver.parse(@json_config_with_cacheTtl)
+    expect(resolver.cache.size).to eq(0)
+    schema = resolver.lookup_schema(@json_config_with_cacheTtl[:schema])
+    expect(resolver.cache.size).to eq(1)
+    sleep resolver.cacheTtl
+    # expect that next request after cacheTtl is undertaken and cache is invalidated
+    schema = resolver.lookup_schema(@json_config_with_cacheTtl[:schema])
+    expect(schema["self"]["version"]).to eq("1-0-2")
+    expect(resolver.cache.size).to eq(0)
```
Comment (chuwy) [inline spec/iglu-client/resolver_spec.rb]: I commented above that this is not correct cache-implementation. Without this test I would not notice it - one more advantage of good test-coverage - you show your expectations to other developers.

File: lib/iglu-client/resolver.rb [hunk 11]
```diff
@@ -50,14 +52,24 @@ def lookup_schema(schema_key)
...
       else
         if cache_result.is_a?(Registries::ResolverError)
           raise cache_result
         else
-          cache_result
+          if cache_result.is_a? (Array)
+              store_time = cache_result[1]
+              time_diff = (fetch_time - store_time).round
+              if time_diff >= @cacheTtl
+                 @cache.delete(schema_key)
```
Comment (chuwy) [inline lib/iglu-client/resolver.rb]: We need not just delete cached item, but replace it with new.

Whole point of cache is that your cached item *can* be outdated after TTL and therefore you need to *refresh* it. In this implementation it is *always* outdated after TTL.

File: spec/iglu-client/resolver_spec.rb [hunk 12]
```diff
@@ -22,14 +22,48 @@
...
+    resolver = Iglu::Resolver.parse(@json_config_with_cacheTtl)
+    expect(resolver.registries.size).to eq(2)
+    expect(resolver.cacheTtl).to eq(5)
+    expect(resolver.registries[1].config.name).to eq("Iglu Central")
+  end
+
+  it 'correctly parses a standard resolver configuration containing cacheTtl (null)' do
+    resolver = Iglu::Resolver.parse(@json_config_with_null_cacheTtl)
+    expect(resolver.registries.size).to eq(2)
+    expect(resolver.cacheTtl).to eq(60)
```
Comment (chuwy) [inline spec/iglu-client/resolver_spec.rb]: Also commented above - we should not have default `cacheTtl`.

File: spec/iglu-client/schema_key_spec.rb [hunk 13]
```diff
@@ -38,15 +38,31 @@
...
+  it 'throws exception on an incorrect SchemaVer (with lower case letters)' do
+    expect { Iglu::SchemaVer.parse_schemaver("a-b-c") }.to raise_error(Iglu::IgluError)
+  end
+
+  it 'throws exception on an incorrect SchemaVer (with upper case letters)' do
+    expect { Iglu::SchemaVer.parse_schemaver("A-B-C") }.to raise_error(Iglu::IgluError)
+  end
+
+  it 'throws exception on an incorrect SchemaVer (dot formatted)' do
+    expect { Iglu::SchemaVer.parse_schemaver("2.0.3") }.to raise_error(Iglu::IgluError)
```
Comment (chuwy) [inline spec/iglu-client/schema_key_spec.rb]: Let's also add few expectations for multiple-digits versions. They're correct, e.g. `10-0-11` should be correct, `10-a-1` should be not correct.

**Review by chuwy — CHANGES_REQUESTED** [review]
Ok, nice. We need one more iteration fixing cache implementation and hopefully that one missing unit-test.

File: lib/iglu-client/resolver.rb (hunk 6 above)
Comment (oguzhanunlu) [reply lib/iglu-client/resolver.rb]: It actually returns schema, not array but I made a mistake here, thanks for noticing!

**Comment by alexanderdean** [comment]
Hey @chuwy - I am working on a "materiality clause" for when a release needs a blog post or not - but definitely this one doesn't need a blog post. Just put the release paragraphs into the Discourse release announcement.

**Comment by chuwy** [comment]
Thanks @alexanderdean! I agree.

Hey @oguzhanunlu, are you sure this PR refers to branch in our repository? I still see `oguzhanunlu:release/0.2.0` in header. If this is "our" branch, then please proceed to release process as we discussed in chat - I'm unlocking master branch.

**Comment by oguzhanunlu** [comment]
Closing since this PR refers to my fork's branch, will open a new PR now. 

Edit: https://github.com/snowplow/iglu-ruby-client/pull/17

**Pull request closed without merging**